geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/kryo/SimpleFeatureSerializer.scala

```diff
@@ -0,0 +1,346 @@
+
+  override def read(kryo: Kryo, input: Input, typ: Class[KryoFeatureId]): KryoFeatureId = {
+    input.readInt(true) // discard version info, currently only one version
+    KryoFeatureId(input.readString())
+  }
+}
+
+/**
+ * Kryo serialization implementation for simple features - provides transformation during read
+ *
+ * @param sft
+ * @param decodeAs
+ */
+class TransformingSimpleFeatureSerializer(sft: SimpleFeatureType, decodeAs: SimpleFeatureType)
+    extends SimpleFeatureSerializer(sft) {
+
+  val transformDecodings = decodings.map {
+    case (decode, i) => (decode, decodeAs.indexOf(sft.getDescriptor(i).getLocalName))
+  }
+
+  override def read(kryo: Kryo, input: Input, typ: Class[SimpleFeature]): SimpleFeature = {
+    input.readInt(true) // discard version info, currently only one version
+    val id = input.readString()
+    val values = Array.ofDim[AnyRef](decodeAs.getAttributeCount)
+
+    transformDecodings.foreach { case (decode, i) =>
+      if (i == -1) decode(input) else values(i) = decode(input)
+    }
+    new KryoSimpleFeature(id, decodeAs, values)
+  }
+}
+
+object SimpleFeatureSerializer {
+
+  import scala.collection.mutable.Map
+
+  val VERSION = 0
+
+  val NULL_BYTE     = 0.asInstanceOf[Byte]
+  val NON_NULL_BYTE = 1.asInstanceOf[Byte]
+
+  type Encoding = (Output, SimpleFeature) => Unit
+  type Decoding = ((Input) => AnyRef, Int)
+
+  // encodings are cached per-thread to avoid synchronization issues
+  // we use soft references to allow garbage collection as needed
+  private val encodingsCache = new ThreadLocal[Map[String, SoftReference[Seq[Encoding]]]] {
+    override def initialValue = Map.empty[String, SoftReference[Seq[Encoding]]]
+  }
+  private val decodingsCache = new ThreadLocal[Map[String, SoftReference[Seq[Decoding]]]] {
+    override def initialValue = Map.empty[String, SoftReference[Seq[Decoding]]]
+  }
+
+  /**
+   * Gets a seq of functions to encode the attributes of simple feature
+   *
+   * @param sft
+   * @return
+   */
+  def sftEncodings(sft: SimpleFeatureType): Seq[Encoding] =
+    encodingsCache.get.get(sft.toString).flatMap(_.get) match {
+      case Some(functions) => functions
+      case None =>
+        val encodings = sft.getAttributeDescriptors.zipWithIndex.map { case (d, i) =>
+          val encode = matchEncode(d.getType.getBinding, d.getUserData)
+          (out: Output, sf: SimpleFeature) => encode(out, sf.getAttribute(i))
+        }
+        encodingsCache.get.put(sft.toString, new SoftReference(encodings))
+        encodings
+    }
+
+  /**
+   * Finds an encoding function based on the input type
+   *
+   * @param clas
+   * @param metadata
+   * @return
+   */
+  def matchEncode(clas: Class[_], metadata: jMap[AnyRef, AnyRef]): (Output, AnyRef) => Unit = clas match {
+
+    case c if classOf[String].isAssignableFrom(c) =>
+      (out: Output, value: AnyRef) => out.writeString(value.asInstanceOf[String])
+
+    case c if classOf[java.lang.Integer].isAssignableFrom(c) =>
```

Comment:
I wonder if there is a way to cut down on the repetition and/or to share this more broadly.  

If Kryo is better than Avro, it might be good to discuss serialization options with the raster team.